src/app/model/classes/administrator.ts

```diff
@@ -0,0 +1,9 @@
+export interface Administrator {
+  email: string;
```

Comment:
non si può usare l'id di un utente invece che una email? in questo modo qualcuno che NON è utente di stalker potrebbe diventare admin

Reply:
ho parlato con Riccardo A di questa cosa e dice che ci occuperemo lato server di verificare se la mail è presente tra gli utenti di Stalker, l'alternativa sarebbe fare una get a un endpoint specifico chiedendo l'Id data la mail

Reply:
non mi convince più di tanto.. la cosa più "giusta" da fare sarebbe avere un'autocomplete nel campo, un po' come quando su github provi a cercare un utente.

bisognerebbe implementare un service + endpoint che ti ritorna la lista (completa di id) di tutti gli utenti collegati all'organizzazione (cosa che dovremo comunque fare) e mantenere un oggetto `connectedUsers: User[]` con questa lista.
poi sul campo del form in cui selezioni di chi vuoi gestire i permessi, scrivi/selezioni un utente e recuperi l'id dall'oggetto.

Reply:
> la cosa più "giusta" da fare sarebbe avere un'autocomplete nel campo, un po' come quando su github provi a cercare un utente

Più che "giusta", questa è la cosa che renderebbe migliore la user experience, ma non è necessariamente più giusta di altre opzioni. Anzi, secondo me non va bene, per una serie di motivi, ma ci torno più sotto.

> la lista (completa di id) di tutti gli utenti collegati all'organizzazione

Non serve a nulla in questo contesto, perchè un admin di un'organizzazione può essere anche qualcuno che non ci è connesso. In particolare, nel caso di un'organizzazione che non richiede l'autenticazione (es: fiera), gli admin non devono mai avere la lista degli utenti connessi, sennò si va contro il punto di non richiedere l'autenticazione. E su sta cosa mi riattacco a quello che ho scritto sopra: se mettiamo l'autocomplete nella scelta del nuovo admin (autocomplete su cosa, poi? sul nome no, perchè ci sono gli omonimi, sulla mail non mi sembra molto sensato, e un nickname non lo abbiamo), finisce che uno si può scorrere l'elenco degli utenti di stalker scorrendo le possibilità nell'autocomplete. Oltretutto, appurato che la scelta del nuovo admin deve essere fra tutti gli utenti, e non solo fra i connessi, servirebbe tenere sto array potenzialmente lunghissimo nella webapp, e stare attenti che rimanga aggiornato nel tempo (aka, richieste http, ripetute nel tempo, con una response gigantesca), andando a rompere tutto quello che mettiamo in piedi per la scalabilità.

Rimango convinto che la cosa più sensata sia usare la mail per identificare gli utenti (visto che è necessariamente univoca, anche se non è chiave primaria del nostro db, ed è la cosa più semplice per l'owner: "Ohi, Toni, dimmi la tua mail, che ti metto admin di sto coso"), e ritornare una response 40X se la mail non c'è nel db.

Reply:
> Più che "giusta", questa è la cosa che renderebbe migliore la user experience, ma non è necessariamente più giusta di altre opzioni. 

è implicitamente più giusta di altre opzioni che hanno UX peggiore

> Non serve a nulla in questo contesto,

il contesto è quello di un istanza del client, cioè tutte le informazioni relative all'organizzazione dell'utente attualmente loggato, che includono la lista degli utenti collegati a tale organizzazione. non è scorretto (anzi, nelle API rest/almost-rest è raccomandato) ottenere dal server TUTTE le informazioni rilevanti, anche se tali informazioni non vengono usate. 

> perchè un admin di un'organizzazione può essere anche qualcuno che non ci è connesso. 

non ricordo se ne abbiamo parlato ma questo non ha senso. ad es. su github non puoi dare permessi di admin su un'organizzazione a qualcuno che non ne fa parte

> In particolare, nel caso di un'organizzazione che non richiede l'autenticazione (es: fiera), gli admin non devono mai avere la lista degli utenti connessi, sennò si va contro il punto di non richiedere l'autenticazione. 

noi non richiediamo l'autenticazione ldap nel caso della fiera, ma vogliamo (anzi dobbiamo) sapere la lista degli utenti collegati a quella specifica organizzazione pubblica.
con riferimento al capitolato:

![image](https://user-images.githubusercontent.com/30255227/80039642-8bc63500-84f8-11ea-949b-7ceccb8d58ae.png)

qua loro intendono "che non preveda autenticazione" ma è sottointeso "autenticazione LDAP". Il layer di "login utente" l'abbiamo aggiunto noi ma è completamente separato da ciò; e ci serve obbligatoriamente per via del GDPR. e inoltre questa cosa aggiunge un livello di complessità ulteriore e non necessario perché stai implicitamente introducendo un terzo tier di organizzazione (pubblica senza lista collegati, pubblica con lista collegati e privata con auth ldap)

> autocomplete su cosa, poi?

`Nome Cognome`

> sul nome no, perchè ci sono gli omonimi, sulla mail non mi sembra molto sensato

`Nome Cognome (email)`

> finisce che uno si può scorrere l'elenco degli utenti di stalker scorrendo le possibilità nell'autocomplete. 

scorretto. si può scorrere l'elenco dei collegati all'organizzazione (cosa che poteva già fare)

> Oltretutto, appurato che la scelta del nuovo admin deve essere fra tutti gli utenti,

appurato in base a un tuo misunderstanding del capitolato e del layer di autenticazione che abbiamo introdotto noi. se facciamo come dici tu o:

- lasciamo che un org tipo fiera tracci tutti gli utenti di stalker (che va contro la premessa da cui siamo partiti, che è "è l'utente che decide chi ha il diritto di tracciarlo)
- introduciamo un terzo tipo di organizzazione (collegato ma l'org non sa che sono collegato, collegato e l'org sa che sono collegato, collegato + ldap) che francamente mi pare insensato e non so da dove ti sia uscita sta cosa.

> stare attenti che rimanga aggiornato nel tempo

no. se tizio apre una pagina sul browser, e caio si collega all'organizzazione dopo che tizio l'ha aperta, non siamo tenuti ad aggiornare la lista. sentiti libero di fare questo tentativo con github, docker hub o qualsiasi altro servizio. e tra l'altro se ne trovi uno che fa così, non vuol dire nulla se non che loro stanno effettivamente facendo

> richieste http, ripetute nel tempo, con una response gigantesca

 se non erro queste richieste una volta venivano fatte con ajax quando si cliccava sul field e si iniziava a scrivere.

> Rimango convinto che la cosa più sensata sia usare la mail per identificare gli utenti

`Nome Cognome (email)`

> ed è la cosa più semplice per l'owner

"aspetta, mrossi@dominio.com è la mail di mario rossi o di mauro rossi?"

> "Ohi, Toni, dimmi la tua mail, che ti metto admin di sto coso")

"ohi toni, dimmi con quale delle tue 5 mail ti sei registrato a sto coso"

Reply:
questa la cosa la metto come issue e la discutiamo meglio lì, più che altro perchè altrimenti non riesco ad andare avanti con il resto